src/wavm.cpp

```diff
@@ -158,27 +84,26 @@ namespace wavm_host_module {
   }
 
 
-  /*
-  useGas
-    Subtracts an amount to the gas counter
-  Parameters
-    amount i64 the amount to subtract to the gas counter
-  Returns
-    nothing
-  */
-  I64 gas = 0;
-  DEFINE_INTRINSIC_FUNCTION_WITH_MEM_AND_TABLE(ethereum, "useGas", void, useGas, I64 amount)
+  DEFINE_INTRINSIC_FUNCTION(ethereum, "call", U32, call, I64 gas, U32 addressOffset, U32 valueOffset, U32 dataOffset, U32 dataLength)
   {
-    printf("useGas  amount=%lu\n",amount);
+    printf("call  gas=%lu  addressOffset=%i  valueOffset=%i  dataOffset=%i  dataLength=%i\n", gas, addressOffset, valueOffset, dataOffset, dataLength);
 
-    try{
-      interface->eeiUseGas(gas);
-    }
-    catch(hera::HeraException e){
-      throwException(Runtime::Exception::calledAbortType);
-    }
+    // save a pointer to the interface before the call
+    WavmEthereumInterface* interface_before_call = interface;
```

Comment:
I think instead of this `interface` should be a `std::stack` where the context is pushed before call, popped after and used with `interface.top()` in each EEI host function.

Reply:
Well actually it should not be pushed/popped in call, but in `wavm_execute`, the entry point to this.

Reply:
You are correct. `interface` is a global defined just below `namespace wavm_host_module {`. Yes, the old pointer is stored during execution of `eeiCall()` so it can be restored to how it was before the call.

Reply:
Oh, if you back it up, replace and restore the backup in the end that will work too. I don't see the full picture, but I feel more comfortable with putting such magic features in execute() rather than in eeiCall().